File: answers/other/answers/leetcode/Perfect Squares.cpp
//@type  Dynamic Programming Breadth-first Search Math
//@result 600 / 600 test cases passed. Status: Accepted Runtime: 404 ms Submitted: 0 minutes ago

class Solution {
public:
    int numSquares(int n) {
		vector<int> square_list(n + 1, INT_MAX);
		square_list[0] = 0;
        for (int i = 0; i <= n; ++ i) {
			for (int j = 1; i + j * j <= n; ++ j) {
				square_list[i + j * j] = min(square_list[i + j * j], square_list[i] + 1);
			}
		}
		return square_list[n];
    }
};

File: answers/leetcode/Perfect Squares/Perfect Squares.java
class Solution {
    
    private int previous = 0;
    private int[] squares = new int[1000000];
    public int numSquares(int n) {
        if (this.squares[n] == 0) {
            for (int i = previous; i <= n; i++) {
                if (i == 0 || this.squares[i] > 0) {
                    for (int j = 0; j * j <= n; j++) {
                        int num = i + j * j;
                        if (num > 0 && this.squares[num] == 0) {
                            this.squares[num] = this.squares[i] + 1;
                        } else {
                            this.squares[num] = Math.min(this.squares[num], this.squares[i] + 1);
                        }
                    }
                }
            }
        }
        return this.squares[n];
    }
}